File: aoj/volume100/ruby/A10015.rb
n           = gets.to_i
exist_cards = $<.map(&:chomp)

suits = %w[S H C D]
ranks = [*1..13]
all_cards     = suits.product(ranks).map{|suit, rank| "#{suit} #{rank}"}
missing_cards = all_cards - exist_cards

puts missing_cards

File: aoj/volume100/haskell/A10015.hs
import Control.Applicative
import Data.List

main = do
  nCards     <- getLine
  existCards <- lines <$> getContents
  mapM_ putStrLn
    $ allCards \\ existCards

allCards =
  cards
  where
    suits = ["S", "H", "C", "D"]
    ranks = [1..13]
    cards = [toCard suit rank | suit <- suits, rank <- ranks]

toCard suit rank =
  suit ++ " " ++ show rank
